modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl2Unfair.java

```diff
@@ -0,0 +1,720 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.datastructures;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.cache.processor.EntryProcessorResult;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteInterruptedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.processors.cache.GridCacheContext;
+import org.apache.ignite.internal.processors.cache.GridCacheIdMessage;
+import org.apache.ignite.internal.processors.cache.IgniteInternalCache;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteInClosure;
+import org.apache.ignite.plugin.extensions.communication.MessageReader;
+import org.apache.ignite.plugin.extensions.communication.MessageWriter;
+
+import static org.apache.ignite.internal.managers.communication.GridIoPolicy.P2P_POOL;
+
+/**
+ * Reentrant lock unfair implementation based on IgniteCache.invoke.
+ */
+public final class GridCacheLockImpl2Unfair extends GridCacheLockEx2 {
+    /** Reentrant lock name. */
+    private final String name;
+
+    /** Cache context. */
+    private final GridCacheContext<GridCacheInternalKey, GridCacheLockState2Base<UUID>> ctx;
+
+    /** Internal synchronization object. */
+    private final LocalSync sync;
+
+    /** */
+    private final GridCacheInternalKey key;
+
+    /** */
+    private final IgniteInClosure<GridCacheIdMessage> releaser;
+
+    /**
+     * Constructor.
+     *
+     * @param name Reentrant lock name.
+     * @param key Reentrant lock key.
+     * @param lockView Reentrant lock projection.
+     */
+    public GridCacheLockImpl2Unfair(String name, GridCacheInternalKey key,
+        IgniteInternalCache<GridCacheInternalKey, GridCacheLockState2Base<UUID>> lockView) {
+        assert name != null;
+        assert key != null;
+        assert lockView != null;
+
+        this.name = name;
+        this.key = key;
+        this.ctx = lockView.context();
+
+        /* Logger. */
+        IgniteLogger log = ctx.logger(getClass());
+
+        // final for passing into anonymous classes.
```

Comment:
do we need to write it explicitly ? it just a java specification rule

Reply:
In Java 8 final is not necessary. So in future, we can remove final :)

Reply:
fixed